source/date_time.c

```diff
@@ -0,0 +1,826 @@
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/common/date_time.h>
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/clock.h>
+#include <aws/common/time.h>
+
+#include <ctype.h>
+#include <math.h>
+
+static const char *RFC822_DATE_FORMAT_STR_MINUS_Z = "%a, %d %b %Y %H:%M:%S GMT";
+static const char *RFC822_DATE_FORMAT_STR_WITH_Z = "%a, %d %b %Y %H:%M:%S %Z";
+static const char *ISO_8601_LONG_DATE_FORMAT_STR = "%Y-%m-%dT%H:%M:%SZ";
+
+static int get_week_day_number_from_str(const char *time_string, size_t startIndex, size_t stopIndex) {
+    if (stopIndex - startIndex < 3) {
+        return -1;
+    }
+
+    size_t index = startIndex;
+
+    char c = time_string[index];
+    char next = 0;
+
+    /* it's ugly but this should compile down to EXACTLY 3 comparisons and no memory allocations */
```

Comment:
Consider using a lexer generator, like [re2c](http://re2c.org/).

You could also use a single switch table by doing:

    switch ((int)(uint8_t)tolower(time_string[0]) | ((int)(uint8_t)tolower(time_string[1]) << 8) | ((int)(uint8_t)tolower(time_string[2]) << 16)

Reply:
I don't want to pull in a dependency for this,
your second suggestion looks promising though.